### PR TITLE
Remove implicit casts of Fiber to bool

### PR DESCRIPTION
**Pull request by sophiebits**
In a couple places I used `==` weak equals -- particularly, element._owner (either null from ReactCurrentOwner, or undefined if it's not present at all), event._dispatchInstances from event system where typing is sloppier, and ChangeEventPlugin which mixes null/undefined. Everywhere else I did strict and believe that it's safe.

Methodology: run tests with this patch and fix all errors.

```diff
diff --git a/scripts/jest/environment.js b/scripts/jest/environment.js
index 71323d7de..96e52aac7 100644
--- a/scripts/jest/environment.js
+++ b/scripts/jest/environment.js
@@ -15,3 +15,15 @@ global.requestIdleCallback = function(callback) {
     });
   });
 };
+
+eval(
+  `
+global.checkBool = function(x) {
+  if (typeof x === 'object' && x !== null && 'stateNode' in x) {
+    //console.log('xyz', new Error().stack);
+    throw new Error('Cast of Fiber to bool detected');
+  }
+  return x;
+};
+`
+);
diff --git a/scripts/jest/preprocessor.js b/scripts/jest/preprocessor.js
index 7727b1b2d..022f6d08c 100644
--- a/scripts/jest/preprocessor.js
+++ b/scripts/jest/preprocessor.js
@@ -53,6 +53,39 @@ var babelOptions = {
     // into ReactART builds that include JSX.
     // TODO: I have not verified that this actually works.
     require.resolve('babel-plugin-transform-react-jsx-source'),
+    function(babel) {
+      const {types: t} = babel;
+
+      return {
+        name: 'check-bool', // not required
+        visitor: {
+          Conditional(path) {
+            path.node.test = t.CallExpression(t.identifier('checkBool'), [
+              path.node.test,
+            ]);
+          },
+          UnaryExpression(path) {
+            if (path.node.operator === '!') {
+              path.node.argument = t.CallExpression(t.identifier('checkBool'), [
+                path.node.argument,
+              ]);
+            }
+          },
+          BinaryExpression(path) {
+            if (path.node.operator === '||' || path.node.operator === '&&') {
+              path.node.left = t.CallExpression(t.identifier('checkBool'), [
+                path.node.left,
+              ]);
+              path.node.right = t.CallExpression(t.identifier('checkBool'), [
+                path.node.right,
+              ]);
+            }
+          },
+        },
+      };
+    },
   ],
   retainLines: true,
 };
```

### CODE REVIEW

File: src/isomorphic/classic/element/ReactElementValidator.js [hunk 1]
```diff
@@ -63,7 +63,7 @@ var ITERATOR_SYMBOL = typeof Symbol === 'function' && Symbol.iterator;
 var FAUX_ITERATOR_SYMBOL = '@@iterator'; // Before Symbol spec.
 
 function getDeclarationErrorAddendum() {
-  if (ReactCurrentOwner.current) {
+  if (ReactCurrentOwner.current !== null) {
     var name = getComponentName(ReactCurrentOwner.current);
     if (name) {
```
Comment (qicaihunanu) [inline src/isomorphic/classic/element/ReactElementValidator.js]: dbgddbdbg

**Comment by gaearon** [comment]
What's the motivation? Performance or correctness?

**Comment by sophiebits** [comment]
Motivation is "I thought we already did it this way", lol. Both performance and precision I guess.

**Comment by sophiebits** [comment]
We could adapt my script to warn on all non-bools if we want to go a little extreme about it.

**Comment by sophiebits** [comment]
Will land after 16.

File: src/renderers/dom/fiber/ReactDOMFiberEntry.js [hunk 2]
```diff
@@ -716,7 +716,7 @@ var ReactDOMFiber = {
       if (__DEV__) {
         const rootEl = getReactRootElementInContainer(container);
         const hasNonRootReactChild = !!(rootEl &&
```
Comment (qicaihunanu) [inline src/renderers/dom/fiber/ReactDOMFiberEntry.js]: 1

File: src/renderers/dom/test/ReactTestUtilsEntry.js [hunk 3]
```diff
@@ -117,16 +114,16 @@ function findAllInRenderedFiberTreeInternal(fiber, test) {
         ret.push(publicInst);
       }
     }
-    if (node.child) {
+    if (node.child !== null) {
```
Comment (qicaihunanu) [inline src/renderers/dom/test/ReactTestUtilsEntry.js]: 1

**Comment by gaearon** [comment]
Closing as stale.

**Pull request closed without merging**